File: solutions/python/math-geometry/zigzag-conversion.py
# LeetCode 6 — Zigzag Conversion (Medium)
# Category: Math & Geometry · Approach: Row Simulation
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/zigzag-conversion/

def convert(s: str, num_rows: int) -> str:
    if num_rows == 1:
        return s
    rows = [""] * num_rows
    row, step = 0, 1
    for c in s:
        rows[row] += c
        if row == 0:
            step = 1
        elif row == num_rows - 1:
            step = -1
        row += step
    return "".join(rows)

File: solutions/typescript/math-geometry/zigzag-conversion.ts
// LeetCode 6 — Zigzag Conversion (Medium)
// Category: Math & Geometry · Approach: Row Simulation
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/zigzag-conversion/

function convert(s: string, numRows: number): string {
  if (numRows === 1) return s;
  const rows = new Array<string>(numRows).fill("");
  let row = 0, step = 1;
  for (const c of s) {
    rows[row] += c;
    if (row === 0) step = 1;
    else if (row === numRows - 1) step = -1;
    row += step;
  }
  return rows.join("");
}
